# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: devtools/server/actors/watcher/WatcherRegistry.jsm
# Commit: 895cf08432c4
# Full Hash: 895cf08432c40fe371ff83e846cb161430c6be4b
# Author: Nicolas Chevobbe <nchevobbe@mozilla.com>
# Date: 2020-10-27 04:41:26
# Regressor Bug: 1633712
# File Overlap Count: 1
# Description:
#   Bug 1633712 - [devtools] Create WorkerTargets as soon as possible. r=ochameau,devtools-backward-compat-reviewers.
#   
#   This patch adds support for dedicated worker targets in the Watcher actor.
#   Shared and Service workers are not handled yet.
#   
# ==============================================================================

diff -r 2ce79a9ed2d5 -r 895cf08432c4 devtools/server/actors/watcher/WatcherRegistry.jsm
--- a/devtools/server/actors/watcher/WatcherRegistry.jsm	Mon Oct 26 12:52:26 2020 +0000
+++ b/devtools/server/actors/watcher/WatcherRegistry.jsm	Mon Oct 26 12:53:43 2020 +0000
@@ -294,31 +294,49 @@
  * Register the JSWindowActor pair "DevToolsFrame".
  *
  * We should call this method before we try to use this JS Window Actor from the parent process
- * via WindowGlobal.getActor("DevToolsFrame").
+ * (via `WindowGlobal.getActor("DevToolsFrame")` or `WindowGlobal.getActor("DevToolsWorker")`).
  * Also, registering it will automatically force spawing the content process JSWindow Actor
  * anytime a new document is opened (via DOMWindowCreated event).
  */
+
+const JSWindowActorsConfig = {
+  DevToolsFrame: {
+    parent: {
+      moduleURI:
+        "resource://devtools/server/connectors/js-window-actor/DevToolsFrameParent.jsm",
+    },
+    child: {
+      moduleURI:
+        "resource://devtools/server/connectors/js-window-actor/DevToolsFrameChild.jsm",
+      events: {
+        DOMWindowCreated: {},
+      },
+    },
+    allFrames: true,
+  },
+  DevToolsWorker: {
+    parent: {
+      moduleURI:
+        "resource://devtools/server/connectors/js-window-actor/DevToolsWorkerParent.jsm",
+    },
+    child: {
+      moduleURI:
+        "resource://devtools/server/connectors/js-window-actor/DevToolsWorkerChild.jsm",
+      events: {
+        DOMWindowCreated: {},
+      },
+    },
+    allFrames: true,
+  },
+};
+
 function registerJSWindowActor() {
   if (isJSWindowActorRegistered) {
     return;
   }
   isJSWindowActorRegistered = true;
-  ActorManagerParent.addJSWindowActors({
-    DevToolsFrame: {
-      parent: {
-        moduleURI:
-          "resource://devtools/server/connectors/js-window-actor/DevToolsFrameParent.jsm",
-      },
-      child: {
-        moduleURI:
-          "resource://devtools/server/connectors/js-window-actor/DevToolsFrameChild.jsm",
-        events: {
-          DOMWindowCreated: {},
-        },
-      },
-      allFrames: true,
-    },
-  });
+  ActorManagerParent.addJSWindowActors(JSWindowActorsConfig);
+
   // Force the immediate activation of this JSWindow Actor, so that we can immediately
   // use the JSWindowActor, from the same event loop.
   ActorManagerParent.flush();
@@ -329,6 +347,9 @@
     return;
   }
   isJSWindowActorRegistered = false;
-  // ActorManagerParent doesn't expose a "removeActors" method, but it would be equivalent to that:
-  ChromeUtils.unregisterWindowActor("DevToolsFrame");
+
+  for (const JSWindowActorName of Object.keys(JSWindowActorsConfig)) {
+    // ActorManagerParent doesn't expose a "removeActors" method, but it would be equivalent to that:
+    ChromeUtils.unregisterWindowActor(JSWindowActorName);
+  }
 }